nrf52-hal-common/src/saadc.rs

```diff
@@ -0,0 +1,122 @@
+//use crate::gpio::{Floating, Input, Pin};
+use crate::{
+    gpio::{Floating, Input},
+    target::{saadc, SAADC},
+};
+use core::{
+    ops::Deref,
+    sync::atomic::{compiler_fence, Ordering::SeqCst},
+};
+use embedded_hal::adc::{Channel, OneShot};
+
+// Only 1 channel is allowed right now, a discussion needs to be had as to how
+// multiple channels should work (See SCAN_MODE)
+
+pub trait SaadcExt: Deref<Target = saadc::RegisterBlock> + Sized {
+    fn constrain(self) -> Saadc;
+}
+
+pub struct Saadc(SAADC);
+
+impl Saadc {
+    pub fn new(saadc: SAADC) -> Self {
+        saadc.enable.write(|w| w.enable().enabled());
+        saadc.resolution.write(|w| w.val()._12bit());
+        saadc.oversample.write(|w| w.oversample().bypass());
+        saadc.samplerate.write(|w| w.mode().task());
+
+        saadc.ch[0].config.write(|w| {
+            w.refsel()
+                .internal()
+                .gain()
+                .gain1_6()
+                .tacq()
+                ._20us()
+                .mode()
+                .se()
+                .resp()
+                .bypass()
+                .resn()
+                .bypass()
+                .burst()
+                .enabled()
+        });
+        saadc.ch[0].pseln.write(|w| w.pseln().nc());
+
+        // Calibrate
+        saadc.tasks_calibrateoffset.write(|w| unsafe { w.bits(1) });
+        while saadc.events_calibratedone.read().bits() == 0 {}
+
+        Saadc(saadc)
+    }
+}
+
+impl<PIN> OneShot<Saadc, u16, PIN> for Saadc
+where
+    PIN: Channel<Saadc, ID = u8>,
+{
+    type Error = ();
+    fn read(&mut self, _pin: &mut PIN) -> nb::Result<u16, Self::Error> {
+        match PIN::channel() {
+            0 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input0()),
+            1 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input1()),
+            2 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input2()),
+            3 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input3()),
+            4 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input4()),
+            5 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input5()),
+            6 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input6()),
+            7 => self.0.ch[0].pselp.write(|w| w.pselp().analog_input7()),
+            // This can never happen as if another pin was used, there would be a compile time error
+            _ => {
+                return Err(nb::Error::Other(()));
```

Comment:
If you're 100% sure about this, you can use https://doc.rust-lang.org/core/hint/fn.unreachable_unchecked.html, which will prevent the compiler from even generating code for this - though the tradeoff is that you get UB if you are wrong :)

Reply:
for PIN::channel to not return a value from 0 to 7 someone would have to modify the channel_mappings macro in some way, and not reflect this back here. Nobody *should* modify it, as those mappings are fixed in hardware for all 3 supported variants, and is unlikely to change even for future variants.

I'm not saying I'm 100% sure, I'm saying that if written *correctly*, this will never happen. Of course if a PR or commit slips through that changes this in some way, then it could be a real problem. I think this is unlikely to happen as there will neither be a need to modify it, nor any future changes that would necessitate modifying it, PLUS the fact that seeing the `unreachable_unchecked()` and comment would alert any future reviewer to pay close attention.

So on that basis I have made the change.